src/io/pipe.cr

```diff
@@ -0,0 +1,81 @@
+require "deque"
+
+class IO::Pipe < IO
+  getter reader, writer
+
+  def initialize(@read_blocking = false, @write_blocking = false, initial_capacity : Int32 = 8192)
+    @buffer = Deque(UInt8).new(initial_capacity: initial_capacity)
+    # TODO: Mutex for atomic reads/writes?
+
+    @reader = uninitialized Reader
+    @writer = uninitialized Writer
+
+    @reader = Reader.new(self)
+    @writer = Writer.new(self)
+  end
+
+  def read(slice : Bytes)
+    count = 0
+    index = 0
+
+    while index < slice.size
+      # TODO: Implement read blocking
+      if byte = @buffer.shift?
+        slice[index] = byte
+        count = (index += 1)
+      else
+        break
+      end
+    end
+    count
+  end
+
+  def write(slice : Bytes) : Nil
+    # TODO: Implement write blocking
+    slice.each do |byte|
+      @buffer << byte
+    end
+  end
+
+  def peek
+    # TODO: Figure out what `closed?` even means here
+    return if closed?
+
+    bytes = Bytes.new(@buffer.size)
+    @buffer.each_with_index do |byte, index|
+      bytes[index] = byte
+    end
```

Comment:
This make no sense. Since when do `peek` return a copy of the whole buffer?

(and it also needs thread safety)

Reply:
> This make no sense. Since when do `peek` return a copy of the whole buffer?

This is a draft and I mentioned in the PR description that this is still very much a WIP.

> (and it also needs thread safety)

Please see [this comment](https://github.com/crystal-lang/crystal/issues/12238#issuecomment-1180331764).